spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binder/AbstractMessageChannelBinder.java

```diff
@@ -258,6 +277,55 @@ protected abstract MessageProducer createConsumerEndpoint(ConsumerDestination de
 	protected void afterUnbindConsumer(ConsumerDestination destination, String group, C consumerProperties) {
 	}
 
+	/**
+	 * Build an errorChannelRecoverer that writes to a pub/sub channel for the destination.
+	 * @param destination the destination.
+	 * @param extendedConsumerProperties the properties.
+	 * @param errorMessageStrategy
+	 * @return the recoverer.
+	 */
+	protected ErrorMessageSendingRecoverer buildRecoveryCallback(ConsumerDestination destination, String group,
+			ExtendedConsumerProperties<?> extendedConsumerProperties, ErrorMessageStrategy errorMessageStrategy) {
+		String errorChannelName = getErrorChannelName(destination, group, extendedConsumerProperties);
+		SubscribableChannel errorChannel;
+		ConfigurableListableBeanFactory beanFactory = getApplicationContext().getBeanFactory();
+		if (getApplicationContext().containsBean(errorChannelName)) {
+			errorChannel = getApplicationContext().getBean(errorChannelName, SubscribableChannel.class);
+		}
+		else {
+			errorChannel = new PublishSubscribeChannel();
+			beanFactory.registerSingleton(errorChannelName, errorChannel);
+			beanFactory.initializeBean(errorChannel, errorChannelName);
+		}
+		ErrorMessageSendingRecoverer recoverer;
+		if (errorMessageStrategy == null) {
+			recoverer = new ErrorMessageSendingRecoverer(errorChannel);
+		}
+		else {
+			recoverer = new ErrorMessageSendingRecoverer(errorChannel, errorMessageStrategy);
+		}
+		String recovererBeanName = getErrorRecovererName(destination, group, extendedConsumerProperties);
+		beanFactory.registerSingleton(recovererBeanName, recoverer);
+		beanFactory.initializeBean(recoverer, recovererBeanName);
+		//TODO unregister on destroy
+		return recoverer;
+	}
+
+	protected String getErrorChannelName(ConsumerDestination destination, String group,
+			ExtendedConsumerProperties<?> extendedConsumerProperties) {
+		return errorsBaseName(destination, group, extendedConsumerProperties) + ".errors";
+	}
+
+	protected String getErrorRecovererName(ConsumerDestination destination, String group,
+			ExtendedConsumerProperties<?> extendedConsumerProperties) {
+		return errorsBaseName(destination, group, extendedConsumerProperties) + ".recoverer";
+	}
+
+	protected String errorsBaseName(ConsumerDestination destination, String group,
+			ExtendedConsumerProperties<?> extendedConsumerProperties) {
```

Comment:
Does the argument need to be `ExtendedConsumerProperties` or can it be `ConsumerProperties` ? Seems we are just using instanceIndex which is coming from the base class.